### PR TITLE
Remove introspection sources on drop cluster replica

### LINKED ISSUE #13882: Cleanup introspection sources after delete replica / delete cluster
As of now we keep the persisted introspection sources around after a replica has been terminated. We should remove them from the catalog and from persist.

Testing:
```
CREATE CLUSTER c1 REPLICAS (r1 (SIZE '1'), r2 (SIZE '2'));
DROP CLUSTER c1;
\dt mz_catalog.*;
```

Should show only the default cluster sources (with a `_1`) postfix, but no `_2` or `_3` sources.\

Also, if we start removing the sources, we need to ensure that views are correctly handled:
```
CREATE view v1 as select * from mz_arrangement_sharing_internal_3;
DROP CLUSTER REPLICA c1.r2;
```
Must fail (due to indices etc. present)

```
CREATE view v1 as select * from mz_arrangement_sharing_internal_3;
DROP CLUSTER REPLICA c1.r2 CASCADE;
```
Must remove `v1`

### PR DESCRIPTION
**Pull request by lluki**
This PR removes the introspection sources (such as `mz_catalog.mz_peek_active_1`) when a replica is dropped. Dependencies are correctly tracked, if there is for example a view defined on the introspection sources, the `DROP REPLICA` will error. I will add the `DROP CLUSTER REPLICA ... CASCADE` in a later PR, but the cascading behavior can be tested with `DROP CLUSTER ... CASCADE`. A test is included.

Fixes #13882 

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### CODE REVIEW

File: src/sql/src/catalog.rs [hunk 1]
```diff
@@ -268,8 +268,11 @@ pub trait CatalogComputeInstance<'a> {
     /// views) of this cluster.
     fn exports(&self) -> &std::collections::HashSet<GlobalId>;
 
-    /// Returns the set of non-transient indexes on this cluster.
+    /// Returns the set of replicas of this cluster.
     fn replica_names(&self) -> HashSet<&String>;
+
+    /// Returns the set of persisted logs of this
```
Comment (mjibson) [inline src/sql/src/catalog.rs:274]: missing word

File: test/sqllogictest/cluster_log_drop.slt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+statement ok
+DROP CLUSTER REPLICA c1.r1;
+
+query T rowsort
+select COUNT(*) from mz_catalog.mz_sources WHERE name LIKE 'mz_peek_active_%';
+----
+2
+
+statement error
```
Comment (mjibson) [inline test/sqllogictest/cluster_log_drop.slt:38]: All of these `statement error` directives need a regex that is a submatch of the full error. For example:
```
statement error materialized view .+ definition names 2 columns, but materialized view .+ has 1 column
```

File: src/adapter/src/coord/sequencer.rs [hunk 3]
```diff
@@ -1403,7 +1403,19 @@ impl<S: Append + 'static> Coordinator<S> {
                     compute_id: instance.id,
                 });
             }
-            let ids_to_drop: Vec<GlobalId> = instance.exports().iter().cloned().collect();
+
+            let mut ids_to_drop: Vec<GlobalId> = instance.exports().iter().cloned().collect();
+
+            // Determine from the replica which additional items to drop. This is the set
+            // of items depend on the introspection sources. The sources
```
Comment (teskje) [inline src/adapter/src/coord/sequencer.rs:1410]:
```suggestion
            // of items that depend on the introspection sources. The sources
```

File: src/adapter/src/coord/sequencer.rs [hunk 4]
```diff
@@ -1440,13 +1453,29 @@ impl<S: Append + 'static> Coordinator<S> {
             });
             let replica_id = instance.replica_id_by_name[&replica_name];
 
+            // Determine from the replica which additional items to drop. This is the set
+            // of items depend on the introspection sources. The sources
```
Comment (teskje) [inline src/adapter/src/coord/sequencer.rs:1457]:
```suggestion
            // of items that depend on the introspection sources. The sources
```

**Review by teskje — APPROVED** [review]
LGTM!

File: test/sqllogictest/cluster_log_drop.slt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# Ensure correct log creation and destruction on `CREATE CLUSTER`, `DROP CLUSTER` DDL statements.
+
+mode cockroach
+
+## REAL starts here
```
[the remark this answers is not in the document]
Comment (lluki) [reply test/sqllogictest/cluster_log_drop.slt:14]: stray

File: test/sqllogictest/cluster_log_drop.slt (hunk 2 above)
Comment (lluki) [reply test/sqllogictest/cluster_log_drop.slt:38]: added!

**Review by mjibson — APPROVED** [review]

**Pull request merged**